java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java

```diff
@@ -140,22 +140,18 @@ public static int getNullCount(final ArrowBuf validityBuffer, final int valueCou
     }
     int count = 0;
     final int sizeInBytes = getValidityBufferSize(valueCount);
+    int remainder = valueCount % 8;
 
-    for (int i = 0; i < sizeInBytes; ++i) {
-      final byte byteValue = validityBuffer.getByte(i);
-      /* Java uses two's complement binary representation, hence 11111111_b which is -1
-       * when converted to Int will have 32bits set to 1. Masking the MSB and then
-       * adding it back solves the issue.
-       */
-      count += Integer.bitCount(byteValue & 0x7F) - (byteValue >> 7);
+    for (int i = 0; i < sizeInBytes; i++) {
+      byte byteValue = validityBuffer.getByte(i);
+      if (i == sizeInBytes - 1 && remainder != 0) {
+        // making the remaining bits all 1s
+        byte mask = (byte)(0xFF << remainder);
+        byteValue = (byte) (byteValue | mask);
+      }
+      count += Integer.bitCount(byteValue & 0xFF);
```

Comment:
This looks ok by me because it will cast up to an integer and only keep the 8 original bits, I do think we need better test coverage for this though